receiver/mysqlreceiver/scraper.go

```diff
@@ -94,12 +94,12 @@ func (m *mySQLScraper) scrape(context.Context) (pmetric.Metrics, error) {
 	globalStats, err := m.sqlclient.getGlobalStats()
 	if err != nil {
 		m.logger.Error("Failed to fetch global stats", zap.Error(err))
-		return pmetric.Metrics{}, err
+		errs.AddPartial(1, err)
```

Comment:
I see why you had to change this, because this is no longer the last query run in the scraper function.

However, I don't think this pattern is appropriate. If there is an error, we should not iterate over the result (even if we know it is "iterable but empty"). Instead, what we should do in this case is push the group of metrics down into another function. This way, the function runs the query and returns immediately if there is an error, otherwise records the metrics from the result. We can follow this pattern for the two queries you added, and a future PR can extract the globalStats query into the same pattern:

```go
// Pass in 'errs' so that each func can append to it as necessary
m.scrapeTableIoWaitsStats(errs)
m.scrapeIndexIoWaitsStats(errs)

...

func (m *mySQLScraper) scrapeTableIoWaitsStats(errs scrapererror.ScrapeErrors) {
	tableIoWaitsStats, err := m.sqlclient.getTableIoWaitsStats()
	if err != nil {
		m.logger.Error("Failed to fetch table io_waits stats", zap.Error(err))
		errs.AddPartial(8, err)
		return
	}

	for i := 0; i < len(tableIoWaitsStats); i++ {
		...
	}
}
```

Reply:
I did this with newly added metrics and I moved them above scraping global stats.
Additionaly, I changed the early return for global stats:
```
        // collect global status metrics.
	globalStats, err := m.sqlclient.getGlobalStats()
	if err != nil {
		m.logger.Error("Failed to fetch global stats", zap.Error(err))
		errs.AddPartial(1, err)
		return m.mb.Emit(), errs.Combine()
	}
```
I'm not sure if this is what you meant, but I couldn't think of other way to do this without refactoring the global stats scraping and without losing any errors.

Reply:
OK, I changed it so that nothing is changed from the original code and I shall create a followup PR tomorrow.

Reply:
Looks exactly correct except that we still need these new functions to be called before the global stats call. Otherwise, a failure of global stats will cause them not to be called.

In the followup PR, it's ok to order them either way. The point is, they are independent and should not affect each other.

Reply:
The PR: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/14458